File: Python/200.py
# 200. Number of Islands

# Given a 2d grid map of '1's (land) and '0's (water),
#  count the number of islands.
#  An island is surrounded by water and is formed by connecting adjacent lands
#  horizontally or vertically.
#  You may assume all four edges of the grid are all surrounded by water.

# Example 1:

# Input:
# 11110
# 11010
# 11000
# 00000

# Output: 1
# Example 2:

# Input:
# 11000
# 11000
# 00100
# 00011

# Output: 3
from typing import List


class Solution:
    def infect(self, grid: List[List[str]], x: int, y: int):
        if x >= len(grid) or y >= len(
                grid[0]) or x < 0 or y < 0 or grid[x][y] != "1":
            return
        else:
            grid[x][y] = "2"
            self.infect(grid, x - 1, y)
            self.infect(grid, x, y - 1)
            self.infect(grid, x + 1, y)
            self.infect(grid, x, y + 1)

    def numIslands(self, grid: List[List[str]]) -> int:
        number = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    number += 1
                    self.infect(grid, i, j)
        return number


if __name__ == '__main__':
    from util import Test
    t = Test(Solution().numIslands)
    t.equal(3, [["1", "1", "0", "0", "0"], ["1", "1", "0", "0", "0"],
                ["0", "0", "1", "0", "0"], ["0", "0", "0", "1", "1"]])
    t.equal(3, [["1", "0", "1", "1", "0", "1", "1"]])

File: cpluscplus/200.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution
{
public:
    int columns = 0;
    int rows = 0;

    void infect(vector<vector<char>> &grid, int x, int y)
    {
        if (y >= columns || x >= rows || y < 0 || x < 0 || grid[y][x] != '1')
        {
            return;
        }
        grid[y][x] = '2';
        infect(grid, x, y + 1);
        infect(grid, x, y - 1);
        infect(grid, x + 1, y);
        infect(grid, x - 1, y);
    }

    int numIslands(vector<vector<char>> &grid)
    {
        if (grid.size() == 0 || grid[0].size() == 0)
        {
            return 0
        }
        int count = 0;
        columns = grid.size();
        rows = grid[0].size();
        for (int y = 0; y < columns; y++)
        {
            for (int x = 0; x < rows; x++)
            {
                if (grid[y][x] == '1')
                {
                    count++;
                    infect(grid, x, y);
                }
            }
        }
        return count;
    }
};

int main()
{
    vector<vector<char>> grid = {
        {'1', '1', '1', '1', '0'},
        {'1', '1', '0', '1', '0'},
        {'0', '0', '1', '0', '0'},
        {'0', '0', '0', '1', '1'},
    };
    int target = 3;
    Solution s = Solution();
    int answer = s.numIslands(grid);
    cout << answer << endl;
    return 0;
}